contracts/example/SnapshotGate/SnapshotGate.sol

```diff
@@ -0,0 +1,335 @@
+
+            // Balance after the transfer
+            uint256 tokenBalanceAfter = IERC20(_tokenAddress).balanceOf(address(this));
+
+            // Make sure that expected amount of tokens was transferred
+            require(
+                tokenBalanceAfter - tokenBalanceBefore == _amount,
+                "Insufficient value received on transfer to gate"
+            );
+
+            // Approve the protocol to transfer this _amount
+            bool success = IERC20(_tokenAddress).approve(protocol, _amount);
+            require(success, "Unable to approve protocol to transfer");
+        }
+    }
+
+    /**
+     * @notice Check the owned and used amounts for a given holder and snapshot token id
+     *
+     * @param _tokenId - the token id to inspect
+     * @param _holder - the holder address to check the balance of
+     * @return owned - the amount owned
+     * @return used - the amount used so far
+     */
+    function checkSnapshot(uint256 _tokenId, address _holder) external view returns (uint256 owned, uint256 used) {
+        owned = snapshot[_tokenId][_holder];
+        used = committed[_tokenId][_holder];
+    }
+
+    /**
+     * @dev Token owner is conditional
+     *
+     * Ultimately, this contract will always remain the owner of any tokens minted.
+     *
+     * However, in the one narrow use-case that the protocol is calling midway through
+     * an in-flight commitToGated offer transaction, we want to report the buyer as
+     * the token owner.
+     *
+     * This is similar to a "flash loan" that is paid back at the end of the transaction.
+     * We have verified that the buyer has the right to commit to the offer, so we 'loan'
+     * them the token for the duration of the transaction, so that they are able to commit.
+     * Ownership reverts to this contract at the end of the transaction.
+     *
+     * Reverts if:
+     * - tokenId does not exist
+     * - a commitToGatedOffer transaction is in-flight and the tokenId does not match txDetails.tokenId
+     *
+     * @param tokenId - the id of the token to check
+     * @return the address of the owner
+     */
+    function ownerOf(uint256 tokenId) public view virtual override returns (address) {
+        // Make sure token exists
+        require(_exists(tokenId), "ERC721: invalid token ID");
+
+        // Determine who to report as the owner
+        address owner;
+        if (txStatus == TransactionStatus.InTransaction) {
+            // Make sure the token id being queried is correct
+            require(tokenId == txDetails.tokenId, "Condition specifies a different tokenId from the one given");
+            // Report owner as stored buyer if in transaction,
+            owner = txDetails.buyer;
+        } else {
+            // When not in transaction, return actual owner
+            // (which will always be this contract)
+            owner = super.ownerOf(tokenId);
+        }
+
+        return owner;
```

Comment:
I considered this when I wrote it, but this function signature is from the spec, and therefore anyone who might be calling it will not expect a name here. Similarly, I don't change the method's `tokenId` argument to the more appropriate `_tokenId`.